File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.HashMap;
class Solution {
    public int lengthOfLongestSubstring(String str) {
        int s = 0; int e = 0; int n = str.length(); int longestSub = 0;
        HashMap<Character, Integer> count = new HashMap<>();
        while (e < n) {
            count.put(str.charAt(e), count.getOrDefault(str.charAt(e), 0) + 1);

            while (count.get(str.charAt(e)) > 1) {
                count.put(str.charAt(s), count.get(str.charAt(s)) - 1);
                if (count.get(str.charAt(s)) == 0) {
                    count.remove(str.charAt(s));
                }
                s++;
            }
            e++;
            longestSub = Math.max(longestSub, e - s);
        }
        return longestSub;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var map: [Character:Int] = [:], lo: String.Index = s.startIndex, hi: String.Index = lo
        var longestSub: Int = 0
        while hi < s.endIndex {
            map[s[hi], default: 0] += 1
            while map[s[hi]]! > 1 {
                map[s[lo]]! -= 1
                if map[s[lo]] == 0 { map[s[lo]] = nil }
                lo = s.index(after: lo)
            }
            hi = s.index(after: hi)
            longestSub = max(longestSub, s.distance(from: lo, to: hi))
        }
        return longestSub
    }
}